agent/hcp/telemetry/otel_sink.go

```diff
@@ -0,0 +1,189 @@
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Logger")
+	}
+
+	if opts.Reader == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Reader")
+	}
+
+	// Setup OTEL Metrics SDK to aggregate, convert and export metrics periodically.
+	res := resource.NewSchemaless()
+	meterProvider := otelsdk.NewMeterProvider(otelsdk.WithResource(res), otelsdk.WithReader(opts.Reader))
+	meter := meterProvider.Meter("github.com/hashicorp/consul/agent/hcp/telemetry")
+
+	gs := &gaugeStore{
+		store: make(map[string]*gaugeValue, 0),
+		mutex: sync.Mutex{},
+	}
+
+	return &OTELSink{
+		spaceReplacer:        strings.NewReplacer(" ", "_"),
+		logger:               opts.Logger.Named("otel_sink"),
+		meterProvider:        meterProvider,
+		meter:                &meter,
+		mutex:                sync.Mutex{},
+		gaugeStore:           gs,
+		gaugeInstruments:     make(map[string]*instrument.Float64ObservableGauge, 0),
+		counterInstruments:   make(map[string]*instrument.Float64Counter, 0),
+		histogramInstruments: make(map[string]*instrument.Float64Histogram, 0),
+	}, nil
+}
+
+// SetGauge emits a Consul gauge metric.
+func (o *OTELSink) SetGauge(key []string, val float32) {
+	o.SetGaugeWithLabels(key, val, nil)
+}
+
+// AddSample emits a Consul histogram metric.
+func (o *OTELSink) AddSample(key []string, val float32) {
+	o.AddSampleWithLabels(key, val, nil)
+}
+
+// IncrCounter emits a Consul counter metric.
+func (o *OTELSink) IncrCounter(key []string, val float32) {
+	o.IncrCounterWithLabels(key, val, nil)
+}
+
+// AddSampleWithLabels emits a Consul gauge metric that gets
+// registed by an OpenTelemetry Histogram instrument.
+func (o *OTELSink) SetGaugeWithLabels(key []string, val float32, labels []gometrics.Label) {
+	k := o.flattenKey(key)
+
+	// Set value in global Gauge store.
+	o.gaugeStore.Store(k, float64(val), toAttributes(labels))
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	// If instrument does not exist, create it and register callback to emit last value in global Gauge store.
+	if _, ok := o.gaugeInstruments[k]; !ok {
+		inst, err := (*o.meter).Float64ObservableGauge(k, instrument.WithFloat64Callback(o.gaugeStore.gaugeCallback(k)))
+		if err != nil {
+			o.logger.Error("Failed to emit gauge: %w", err)
+			return
+		}
+		o.gaugeInstruments[k] = &inst
+	}
+}
+
+// AddSampleWithLabels emits a Consul sample metric that gets registed by an OpenTelemetry Histogram instrument.
+func (o *OTELSink) AddSampleWithLabels(key []string, val float32, labels []gometrics.Label) {
+	k := o.flattenKey(key)
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	inst, ok := o.histogramInstruments[k]
+	if !ok {
+		histogram, err := (*o.meter).Float64Histogram(k)
+		if err != nil {
+			o.logger.Error("Failed to emit gauge: %w", err)
+			return
+		}
+		inst = &histogram
+		o.histogramInstruments[k] = inst
+	}
+
+	attrs := toAttributes(labels)
+	(*inst).Record(context.TODO(), float64(val), attrs...)
```

Comment:
Should we be using a cancelable context here?

Reply:
Yeah good question, so we have a timeout for the actual net/http request which happens at the export stage. 
In this block of code, the measurement of a metric gets recorded and added to the aggregation. The export happens every 10 seconds, and it's not related to this code block.

From my digging in the SDK, the context is passed here as an early return in case the ctx is cancelled. 
So if the ctx is cancelled, metrics won't get aggregated nor recorded, see source code: 

https://github.com/open-telemetry/opentelemetry-go/blob/24e8d2e5749d22fcd2266eb3e4d35bbe9d42259f/sdk/metric/instrument.go#L190C2-L202

In their usage [example](https://github.com/open-telemetry/opentelemetry-go/blob/24e8d2e5749d22fcd2266eb3e4d35bbe9d42259f/example/prometheus/main.go#L82.) in their official docs, they use a `context.Background` and await interrupt signals / ctx cancellation.

The right thing to do here would be to use a parent context, to prevent metric aggregation in the case where ctx is cancelled. However, we don't have that in the code right now since the go metrics sink methods don't pass in a context 😓 I feel like they should... A timeout could ensure this aggregation doesn't take too long - which it shouldn't. We could also end up missing metrics. In think for that, should we be more data driven and look at performance first to see if it's needed?

Reply:
I was definitely unsure what to do here, so let me know if you have any preferences